location/lbs/activity/audioset/dataprocessing/audio_processing.py

```diff
@@ -0,0 +1,576 @@
+
+ --src_dir is the path to the directory containing the input files
+(audioset csv file ontology.json)
+--dest_dir is the path to the directory containing the output files
+(output csv file)
+--labels included in the AudioSet dataset can be used (i.e "Gunshot, gunfire")
+
+Performance: to confirm the presence of, extract features from, and output to a
+    csv file, takes about 3 hours and ___ minutes for around 20,000 samples.
+"""
+
+from __future__ import unicode_literals
+import youtube_dl
+import csv
+import json
+from pydub import AudioSegment
+import os
+import shutil
+from os import listdir
+from os.path import isfile, isdir, join
+import librosa
+from absl import app
+from absl import flags
+from absl import logging
+import datetime
+import pandas as pd
+import numpy as np
+
+FLAGS = flags.FLAGS
+flags.DEFINE_string('filename', 'balanced_train_segments',
+                    'The csv file of a dataset provided by the audioset')
+flags.DEFINE_multi_string('labels', 'Gunshot, gunfire',
+                          'The set of labels identified by the audioset to '
+                          'treat as positive')
+flags.DEFINE_multi_string('features', 'mfcc',
+                          'The set of features to extract from the audio files')
+flags.DEFINE_bool('redo', False,
+                  'Whether to redownload the YouTube videos in the included'
+                  ' csv file')
+flags.DEFINE_string('src_dir',
+                    'location/lbs/activity/audioset/dataprocessing/example_src_dir',
+                    'The location of the audioset csv file(s) and the '
+                    'ontology.json file')
+flags.DEFINE_string('dest_dir',
+                    'location/lbs/activity/audioset/dataprocessing/example_dest_dir',
+                    'The location of the csv file and downloaded YouTube '
+                    'videos')
+flags.DEFINE_bool('scaled', True, 'Whether to average the extracted features')
+
+
+def parse_metadata(src_dir, filename):
+    """Parses metadata and labels from the audioset csv.
+
+    Iterates through the csv files from audioset and extracts the metadata:
+    video_id, start_time, end_time, and labels
+
+    Args:
+        src_dir: Path to a directory where the audioset csv file is expected to
+            be found.
+        filename: A csv file without the .csv file extension
+
+    Returns:
+        A tuple with the first element being a list of lists containing a
+        video_id, start_time (in seconds), and end_time (in seconds). This list
+        is generated from the inputted csv. The second element is a dictionary
+        with video_id, list of labels key, value pairs
+        For example:
+
+        ([[video_id1, start_time1, end_time1],
+        [video_id2, start_time2, end_time2],
+        [video_id3, start_time3, end_time3]],
+        {video_id1 : [label1, label3],
+        video_id12: [label2]
```

Comment:
This dictionary is starting to look a little complicated. What you can do instead is create a class.

class AudioSetEntry(object):
  def \_\_init\_\_(self, video_id, start_time, end_time, labels):
    ...

And have a dictionary of video ids -> AudioSetEntries that can be referenced later. With dictionaries, you can also use iteritems() which will iterate through each key value pair as tuples.

https://python-reference.readthedocs.io/en/latest/docs/dict/iteritems.html

Reply:
Another option is to use csv.DictReader, which automatically return a dictionary for you, which means you don't need to use a class.

https://docs.python.org/3/library/csv.html#csv.DictReader

Reply:
Because of the labels being wrapped in quotations, I used some extra processing to remove the quotations marks and extract a list of labels. I can create a class, however, to congregate all the metadata in one place.